src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
     {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return ConnectStateMachineAsync(false, connectOpts, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return ConnectStateMachineAsync(true, connectOpts, cancellationToken);
-        }
-
+        Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
         try
         {
             if (_disposed)
             {
                 throw new ObjectDisposedException(nameof(CommandWriter));
             }
 
-            var success = false;
-            try
-            {
-                _protocolWriter.WriteConnect(connectOpts);
-                success = true;
-            }
-            finally
+            PipeWriter bw;
+            lock (_lock)
             {
-                EnqueueCommand(success);
+                if (_pipeWriter == null)
+                    ThrowOnDisconnected();
+                bw = _pipeWriter!;
             }
+
+            _protocolWriter.WriteConnect(bw, connectOpts!);
+            await bw.FlushAsync(cancellationToken).ConfigureAwait(false);
         }
         finally
         {
             _semLock.Release();
+            Interlocked.Add(ref _counter.PendingMessages, -1);
         }
-
-        return ValueTask.CompletedTask;
     }
 
-    public ValueTask PingAsync(PingCommand pingCommand, CancellationToken cancellationToken)
+    public async ValueTask PingAsync(PingCommand pingCommand, CancellationToken cancellationToken)
     {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return PingStateMachineAsync(false, pingCommand, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return PingStateMachineAsync(true, pingCommand, cancellationToken);
-        }
-
+        Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
         try
         {
             if (_disposed)
             {
                 throw new ObjectDisposedException(nameof(CommandWriter));
             }
 
-            var success = false;
-            try
-            {
-                _protocolWriter.WritePing();
-                _enqueuePing(pingCommand);
-                success = true;
-            }
-            finally
+            PipeWriter bw;
+            lock (_lock)
             {
-                EnqueueCommand(success);
+                if (_pipeWriter == null)
+                    ThrowOnDisconnected();
+                bw = _pipeWriter!;
             }
```

Comment:
Given the re-use of this pattern, is there any potential value in us trying to consolidate shared logic of the lock/get (and possibly more) into a method call to lower maintenance concerns? (Maybe not at this stage but though it was worth asking.)